paddle/fluid/operators/cudnn_gru_op.cu.cc

```diff
@@ -0,0 +1,291 @@
+
+    size_t max_len = ctx.Attr<int>("max_len");
+    float dropout_prob = ctx.Attr<float>("dropout_prob");
+    bool is_bidirec = ctx.Attr<bool>("is_bidirec");
+    int input_size = ctx.Attr<int>("input_size");
+    int hidden_size = ctx.Attr<int>("hidden_size");
+    int num_layers = ctx.Attr<int>("num_layers");
+    bool is_test = ctx.Attr<bool>("is_test");
+
+    auto &dev_ctx = ctx.template device_context<platform::CUDADeviceContext>();
+    auto handle = dev_ctx.cudnn_handle();
+    auto *cache_var = ctx.InputVar("Cache");
+    if (!cache_var) {
+      // The RAW type cache variable wouldn't be created and broadcasted on
+      // multi-devices before the first running.
+      // use parent scope to make cache persistable
+      auto *scope = const_cast<framework::Scope *>(ctx.scope().parent());
+      auto cache_var_name = ctx.Inputs("Cache")[0];
+      cache_var = scope->Var(cache_var_name);
+    }
+    CudnnRNNCache *cudnn_rnn_cache = nullptr;
+    T *x_data_padding = nullptr;
+    int max_seq_len = 0;
+    std::vector<int> seq_len;
+    framework::DDim x_dim = x->dims();
+    if (cache_var->IsInitialized()) {
+      cudnn_rnn_cache = const_cast<framework::Variable *>(cache_var)
+                            ->GetMutable<CudnnRNNCache>();
+    } else {
+      cudnn_rnn_cache = const_cast<framework::Variable *>(cache_var)
+                            ->GetMutable<CudnnRNNCache>();
+      std::random_device rnd;
+      int seed = ctx.Attr<int>("seed");
+      if (seed == -1) {
+        seed = rnd();
+      }
+
+      auto input_w_numel = w->numel();
+      int batch_size = 1;
+      // if input not lod tensor
+      if (x->lod().size() == 0) {
+        batch_size = x_dim[1];
+      } else {
+        framework::LoD x_lod = x->lod();
+        batch_size = x_lod[0].size() - 1;
+        max_len = x_dim[0];
+        // if batch size > 1 start batch_size cudnn.
+        if (batch_size > 1) {
+          for (int i = 0; i < x_lod[0].size() - 1; ++i) {
+            int len = x_lod[0][i + 1] - x_lod[0][i];
+            seq_len.push_back(len);
+            if (len > max_seq_len) max_seq_len = len;
+          }
+          max_len = max_seq_len;
+          batch_size = 1;
+        }
+      }
+      cudnn_rnn_cache->init(handle, ctx.GetPlace(), max_len, batch_size,
+                            input_size, hidden_size, num_layers, dropout_prob,
+                            is_bidirec, seed, input_w_numel, true);
+    }
+
+    auto run_seq_len = x->dims()[0];
+    T *out_data = out->mutable_data<T>(ctx.GetPlace());
+    if (seq_len.size() > 0) {
+      int x_padding_size = max_seq_len * x_dim[1];
+      cudaErrCheck(cudaMalloc(reinterpret_cast<void **>(&x_data_padding),
+                              x_padding_size));
+      // do cudnn for each lod, maybe need refine.
+      for (auto &len : seq_len) {
+        int mem_size = len * x_dim[1];
+        cudaErrCheck(cudaMemcpy(x_data_padding, x_data, mem_size * sizeof(T),
```

Comment:
PADDLE_ENFORCE has the check for cuda related call

No need to call the raw cuda check API